### ISSUE TITLE
fuzzer crash in http2_done

### ISSUE THREAD

**Issue by bagder** [curl/curl#5646]
This started with the merge of #5643.

We get the [crash](https://github.com/curl/curl/pull/5644/checks?check_run_id=834806110#step:5:31) with nghttp2 1.41.0:
~~~
==16==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x603000341e92 at pc 0x00000068d457 bp 0x7fff967bf520 sp 0x7fff967bf518
READ of size 1 at 0x603000341e92 thread T0
SCARINESS: 22 (1-byte-read-heap-buffer-overflow-far-from-bounds)
    #0 0x68d456 in nghttp2_session_check_request_allowed /src/nghttp2/lib/nghttp2_session.c:1482:20
    #1 0x5887e5 in Curl_http2_done /src/curl/lib/http2.c:1207:11
    #2 0x57bb4a in Curl_http_done /src/curl/lib/http.c:1537:3
    #3 0x5d3679 in rtsp_done /src/curl/lib/rtsp.c:209:16
    #4 0x5b16b2 in multi_done /src/curl/lib/multi.c:581:14
    #5 0x5b7114 in multi_runsingle /src/curl/lib/multi.c:1630:15
    #6 0x5b5b5a in curl_multi_perform /src/curl/lib/multi.c:2406:14
    #7 0x5516a3 in fuzz_handle_transfer(fuzz_data*) /src/curl_fuzzer/curl_fuzzer.cc:383:5

_Originally posted by @bagder in https://github.com/curl/curl/pull/5643#issuecomment-653596795_

**Comment by bagder** [comment]
I should add that this is the CI-Fuzz job that crashes and the link shows it crashed for an unrelated PR. Reproducing the exact same crash is probably a little cumbersome.

**Comment by laramiel** [comment]
Is there some other signaling that the nghttp2 session has already been closed? I notice that most of the other checks in the method check the stream_id as well.

**Comment by laramiel** [comment]
I spent some time looking at this, but didn't find a codepath that leads to this state.

So instead I reconsidered where I had added the connclose() call:

curl signals that data is available to the nghttp2 protocol library by calling nghttp2_session_mem_recv(). Instead of checking at http_done, instead we can check whether the server has requested a connection be closed after the mem_recv() call, since that's the only place where the state changes.

So I reviewed all the mem_recv() calls. It turns out that there were 3 codepaths, all similar, but also different.
I consolidated them to use the existing h2_process_pending_input() method.
